### PR TITLE
Update for CLAP 1.1.7

### PR DESCRIPTION
**Pull request by robbert-vdh**
This covers the 1.1.2 update through 1.1.7, including the new preset discovery factory. This supersedes #23.

There have been a lot of new additions since 1.1.1 so I'm sure there's at least one mistake in here somewhere.

### CODE REVIEW

File: src/ext/audio_ports_config.rs [hunk 1]
```diff
@@ -42,6 +44,21 @@ pub struct clap_plugin_audio_ports_config {
...
+#[derive(Debug, Copy, Clone)]
+pub struct clap_plugin_audio_ports_config_info {
+    pub current_config: Option<unsafe extern "C" fn(plugin: *const clap_plugin) -> clap_id>,
+    pub get: Option<
+        unsafe extern "C" fn(
+            plugin: *const clap_plugin,
+            config_id: clap_id,
+            port_index: u32,
+            is_input: bool,
+            config: *mut clap_audio_ports_config,
```
Comment (micahrj) [inline src/ext/audio_ports_config.rs]: Should be `info: *mut clap_audio_port_info`: https://github.com/free-audio/clap/blob/065d685d4e9657f0344f350eef748be2b4d8e318/include/clap/ext/audio-ports-config.h#L90

File: src/ext/draft/context_menu.rs [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+        ) -> bool,
+    >,
+    pub can_popup: Option<unsafe extern "C" fn(host: *const clap_host) -> bool>,
+    pub popup: Option<
+        unsafe extern "C" fn(
+            host: *const clap_host,
+            target: *const clap_context_menu_target,
+            screen_index: u32,
+            x: u32,
+            y: u32,
```
Comment (micahrj) [inline src/ext/draft/context_menu.rs]: These should be `i32`: https://github.com/free-audio/clap/blob/065d685d4e9657f0344f350eef748be2b4d8e318/include/clap/ext/draft/context-menu.h#L151-L153

File: src/factory/draft/preset_discovery.rs [hunk 3]
```diff
@@ -0,0 +1,228 @@
...
+    pub uri: *const c_char,
+}
+
+unsafe impl Send for clap_preset_discovery_location {}
+unsafe impl Sync for clap_preset_discovery_location {}
+
+#[repr(C)]
+#[derive(Debug, Copy, Clone)]
+pub struct clap_preset_discovery_soundpack {
+    pub flags: u32,
```
Comment (micahrj) [inline src/factory/draft/preset_discovery.rs]: Should be `u64`: https://github.com/free-audio/clap/blob/065d685d4e9657f0344f350eef748be2b4d8e318/include/clap/factory/draft/preset-discovery.h#L203

File: src/factory/draft/preset_discovery.rs [hunk 4]
```diff
@@ -0,0 +1,228 @@
...
+    pub id: *const c_char,
+    pub name: *const c_char,
+    pub description: *const c_char,
+    pub homepage_url: *const c_char,
+    pub vendor: *const c_char,
+    pub image_url: *const c_char,
```
Comment (micahrj) [inline src/factory/draft/preset_discovery.rs]: Should be `image_uri`: https://github.com/free-audio/clap/blob/065d685d4e9657f0344f350eef748be2b4d8e318/include/clap/factory/draft/preset-discovery.h#L209

**Review by micahrj — CHANGES_REQUESTED** [review]

File: src/factory/draft/preset_discovery.rs (hunk 3 above)
Comment (robbert-vdh) [reply src/factory/draft/preset_discovery.rs]: u32 is actually correct here. The spec is inconvenient and uses both interchangeably. But it's primarily used as u32.

Comment (micahrj) [reply src/factory/draft/preset_discovery.rs]: I don't understand; the struct layouts of `clap_preset_discovery_soundpack` will not match if the CLAP headers use `uint64_t` and `clap-sys` uses `u32`.

Comment (robbert-vdh) [reply src/factory/draft/preset_discovery.rs]: Oops sorry, I opened this on my phone and I thought the comment was about the `u32` type for the `CLAP_PRESET_DISCOVERY_IS_*` flags. Fixed, along with the other errors.

File: src/ext/audio_ports_config.rs [hunk 5]
```diff
@@ -42,6 +45,21 @@ pub struct clap_plugin_audio_ports_config {
...
+#[derive(Debug, Copy, Clone)]
+pub struct clap_plugin_audio_ports_config_info {
+    pub current_config: Option<unsafe extern "C" fn(plugin: *const clap_plugin) -> clap_id>,
+    pub get: Option<
+        unsafe extern "C" fn(
+            plugin: *const clap_plugin,
+            config_id: clap_id,
+            port_index: u32,
+            is_input: bool,
+            config: *mut clap_plugin_audio_ports,
```
Comment (micahrj) [inline src/ext/audio_ports_config.rs]: This is still wrong, it needs to be `clap_audio_port_info`: https://github.com/free-audio/clap/blob/065d685d4e9657f0344f350eef748be2b4d8e318/include/clap/ext/audio-ports-config.h#L90

**Review by micahrj — CHANGES_REQUESTED** [review]

File: src/ext/audio_ports_config.rs (hunk 5 above)
Comment (robbert-vdh) [reply src/ext/audio_ports_config.rs]: Oops, somehow missed that. Fixed.

**Review by micahrj — APPROVED** [review]

**Pull request merged**